### ISSUE TITLE
copilot revision task

### ISSUE THREAD

**Issue by Eupham** [Eupham/KernelDev#81]
###THIS IS LAW!!!!: DO NOT ALLOW YOURSELF TO TRY GETTING OUT OF USING ORIGINAL KERNEL.PY THIS MUST BE USED###
Here’s the crisp target behavior for the party (cocktail_party) task:

Input shape (one [MASKQ]): task details [CLS] this is a [MASK] it has whiskers and meows [SPAN]duck[ES][SPAN]cat[ES][SPAN]dog[ES] [MASKQ]

Token roles & what they may see (query → keys):

[CLS] / prefix tokens (the “task details” prefix): Keys for everyone, but they themselves don’t look forward past the [CLS] boundary (they only see within the prefix; do not claim they see the whole sequence).

Normal context tokens (the sentence with [MASK] in it): Causal within context; cannot attend to any span tokens; may attend to prefix; acts like a standard LM region.

[MASK] (marker only): Same permissions as context (causal, no span access). It is not the scorer.

[SPAN] … [ES] (each candidate “island”): Bidirectional inside the same span only; may attend to [CLS] and [MASKQ]; cannot attend to other spans or the normal context (prevents leakage).

[MASKQ] (one per sequence, placed after all spans): Bidirectional access to all spans + [CLS] (hub/aggregator). For simpler calculations maskq should only see spans. Spans already have access to context and context already has access to cls.

User says this behavior trumps any other conceptions you may derive from the res of this###Think along the lines of Directionality matters. Here are our sections
1 [cls] these tokens are bidirectional and can only see section 1
2 these tokens are causal and may include the [mask] token, they see section 1 and themself causally
3 [span]<span island content 1>[es][span][es] these tokens see bidirectionally and can see themself and section 2. Each island cannot see another island. 3/span islands see context/2.  !Critical! span/islands must not see maskq
4 [maskq] this is our bridge token, it can see all islands at the same time bidirectionally.  islands cannot see maskq
1 cannot see 2-4, 2 cannot see 3-4, 3 cannot see 1 directly (but can see indirectly via 2) and cannot see 4 each island cannot attend to another island, 4 cannot see 1 or 2 directly (but can see indirectly via 3)

Isolation guarantees (no leakage):

Context (incl. [MASK]) never sees span tokens.

A span never sees another span or the context; any cross-span comparison must flow through [MASKQ].  Both must be true.  MASKQ must see all spans and all spans must not see maskq

Scoring & loss:

Pool each span embedding by mean over tokens between [SPAN]..[ES].

Use the [MASKQ] embedding as the query: score_i = h_[MASKQ] · h_span_i.

Cross-entropy over scores with the gold index (which span is correct).

Cardinality/placement constraints:

Exactly one [MASKQ] per sequence, appended after all spans.  Spans must not read maskq

Keep [MASK] in the context where the original span was removed.

Support bidirectional_prefix_len (≥1) so the prefix can be N tokens, but remember: prefix doesn’t look forward.

High-level intent:

Force the model to choose among isolated span candidates using a single global hub ([MASKQ]) that can read all spans (and the task prefix) while forbidding any direct context→span or span→span peeking.  Span MUST NOT read maskq.

Determine areas for additional purging (we should only have 2 tasks and no references to other tasks anywhere, so we should only be referencing teacher forcing and cocktail party). Determine what I need to change to meet my goals for token behavior. In areas where both fwd and bwd behavior is done we need to make sure both areas are accounted for

It seems like we are using 2 methods to mask. One is an attention mask in the model side, and another is a mask on the kernel side. Lets pick the one that's less computationally expensive.

Usernote: We are running this from the pov of entry.py running such as
!python KernelDev/entry.py --nproc_per_node=1 --config config.yaml --epochs 1 --precision bf16
